### PR TITLE
Lazily decode series protobuf.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
The protobuf decoding for series responses runs into a memory issue for many series. The response is only merged and passed through the front end. It is more efficient to decode the protobuf encoded series lazily. And `decode`, `merge`, `encode` benchmark shows the benefit of transcoding the protobuf series message into a JSON response directly.

This change will not impact production code since it only applies to protobuf encoding messaging between the querier and query frontend that must be explicitly enabled.

```
› go test -v -run=^$ -bench "Benchmark_DecodeMergeEncodeCycle" -memprofile memory_base.prof -count=10 ./pkg/querier/queryrange > before.txt
› go test -v -run=^$ -bench "Benchmark_DecodeMergeEncodeCycle" -memprofile memory_base.prof -count=10 ./pkg/querier/queryrange > before.txt
› benchstat before.txt after.txt
before.txt:5: missing iteration count
after.txt:5: missing iteration count
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/querier/queryrange
cpu: AMD Ryzen 7 3700X 8-Core Processor             
                           │  before.txt  │              after.txt              │
                           │    sec/op    │   sec/op     vs base                │
_DecodeMergeEncodeCycle-16   2537.7m ± 2%   934.2m ± 1%  -63.19% (p=0.000 n=10)

                           │  before.txt   │              after.txt               │
                           │     B/op      │     B/op      vs base                │
_DecodeMergeEncodeCycle-16   1723.4Mi ± 0%   641.1Mi ± 0%  -62.80% (p=0.000 n=10)

                           │  before.txt   │              after.txt              │
                           │   allocs/op   │  allocs/op   vs base                │
_DecodeMergeEncodeCycle-16   20240.6k ± 0%   203.0k ± 0%  -99.00% (p=0.000 n=10)

```

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: pkg/querier/queryrange/views.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func WriteSeriesResponseViewJSON(v *MergedSeriesResponseView, w io.Writer) error {
+	_, err := io.WriteString(w, `{ "status": "success", "data": [`)
```
Comment (jeschkies) [inline pkg/querier/queryrange/views.go:193]: Use jsoniter instead to escape strings.

**Comment by jeschkies** [comment]
```
go test -v -run=^$ -bench "Benchmark_DecodeMergeEncodeCycle" -memprofile memory_base.prof -benchtime=10x ./pkg/querier/queryrange > before.txt
go test -v -run=^$ -bench "Benchmark_DecodeMergeEncodeCycle" -memprofile memory.prof -benchtime=10x ./pkg/querier/queryrange > after.txt 
go tool pprof -http :9402 --base memory_base.prof memory.prof
```
image

As can be seen `AsBytesSafe` is allocating the most but the benefit is still big.

File: pkg/querier/queryrange/codec.go [hunk 2]
```diff
@@ -777,14 +785,20 @@ func encodeResponseProtobuf(ctx context.Context, res queryrangebase.Response) (*
 		p.Response = &QueryResponse_Streams{response}
 	case *LokiSeriesResponse:
 		p.Response = &QueryResponse_Series{response}
+	case *MergedSeriesResponseView:
+		mat, err := response.Materialize()
+		if err != nil {
+			return nil, err
```
Comment (dannykopping) [inline pkg/querier/queryrange/codec.go:791]: Wrap this error please

File: pkg/querier/queryrange/views.go [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+	"github.com/richardartoul/molecule"
+	"github.com/richardartoul/molecule/src/codec"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+)
+
+// GetLokiSeriesResponseView returns a view on the series response of a
+// QueryResponse. Returns and error if the message was empty. Note: the method
+// does not verify that the reply is a propberly encoded QueryResponse protobuf.
```
Comment (dannykopping) [inline pkg/querier/queryrange/views.go:19]:
```suggestion
// GetLokiSeriesResponseView returns a view on the series response of a
// QueryResponse. Returns an error if the message was empty. Note: the method
// does not verify that the reply is a properly encoded QueryResponse protobuf.
```

File: pkg/querier/queryrange/views.go [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+func GetLokiSeriesResponseView(data []byte) (view *LokiSeriesResponseView, err error) {
+	b := codec.NewBuffer(data)
+	err = molecule.MessageEach(b, func(fieldNum int32, value molecule.Value) (bool, error) {
+		if fieldNum == 1 {
```
Comment (dannykopping) [inline pkg/querier/queryrange/views.go:23]: Could we get the field index somehow to make it more clear which field we're referring to?

File: pkg/querier/queryrange/views.go [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+			if len(value.Bytes) > 0 {
+				// We might be able to avoid an allocation and
+				// copy here by using value.Bytes
+				data, err = value.AsBytesSafe()
+				if err != nil {
+					return false, err
```
Comment (dannykopping) [inline pkg/querier/queryrange/views.go:29]: Please wrap error

File: pkg/querier/queryrange/views.go [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+// Implement proto.Message
+func (v *LokiSeriesResponseView) Reset()         {}
+func (v *LokiSeriesResponseView) String() string { return "" }
+func (v *LokiSeriesResponseView) ProtoMessage()  {}
+
+// ForEachSeries iterates of the []logproto.SeriesIdentifier slice and pass a
+// view on each identifier to the callback supplied.
+func (v *LokiSeriesResponseView) ForEachSeries(fn func(view *SeriesIdentifierView) error) error {
+	return molecule.MessageEach(codec.NewBuffer(v.buffer), func(fieldNum int32, value molecule.Value) (bool, error) {
+		if fieldNum == 2 {
```
Comment (dannykopping) [inline pkg/querier/queryrange/views.go:67]: Ditto

File: pkg/querier/queryrange/views.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+			pair = pair[:0]
+
+			return true, nil
+		}
+
+		return true, nil
+	})
+}
+
+var sep = string([]byte{'\xff'})
```
Comment (dannykopping) [inline pkg/querier/queryrange/views.go:129]: This is copied from Prometheus, right? We should indicate that

File: pkg/querier/queryrange/views.go [hunk 8]
```diff
@@ -0,0 +1,332 @@
+package queryrange
```
Comment (dannykopping) [inline pkg/querier/queryrange/views.go:1]: Nit: this file is rather large and very complex, we might consider splitting it up.

File: pkg/querier/queryrange/views.go [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+					return false, err
+				}
+				pair = append(pair, s)
+				return true, nil
+			})
+			if err != nil {
+				return false, err
+			}
+
+			// TODO(karsten): check length of pair
```
Comment (dannykopping) [inline pkg/querier/queryrange/views.go:113]: TODO

**Review by dannykopping — COMMENTED** [review]
These results look super impressive @jeschkies! Left mostly nits & small suggestions

File: pkg/querier/queryrange/views.go (hunk 4 above)
Comment (jeschkies) [reply pkg/querier/queryrange/views.go:23]: You mean from the descriptor? Or would it be enough to declare a constant.

File: pkg/querier/queryrange/views.go (hunk 9 above)
Comment (jeschkies) [reply pkg/querier/queryrange/views.go]: Addressed

File: pkg/querier/queryrange/views.go (hunk 5 above)
Comment (jeschkies) [reply pkg/querier/queryrange/views.go]: Addressed

File: pkg/querier/queryrange/views.go (hunk 3 above)
Comment (jeschkies) [reply pkg/querier/queryrange/views.go]: Addressed

File: pkg/querier/queryrange/codec.go (hunk 2 above)
Comment (jeschkies) [reply pkg/querier/queryrange/codec.go:791]: Addressed

File: pkg/querier/queryrange/views.go (hunk 8 above)
Comment (jeschkies) [reply pkg/querier/queryrange/views.go:1]: I thought about this but wanted to make it easy to remove. It's less than 400 lines. `codec.go` is 1300+. I'm happy to split it into each view though.

File: pkg/querier/queryrange/views.go [hunk 10]
```diff
@@ -266,10 +268,14 @@ func (v *MergedSeriesResponseView) Materialize() (*LokiSeriesResponse, error) {
 	mat := &LokiSeriesResponse{}
 	err := v.ForEachUniqueSeries(func(series *SeriesIdentifierView) error {
 		identifier := logproto.SeriesIdentifier{Labels: make(map[string]string)}
-		series.ForEachLabel(func(name, value string) error {
+		err := series.ForEachLabel(func(name, value string) error {
 			identifier.Labels[name] = value
 			return nil
 		})
+		if err != nil {
+			return err
```
Comment (dannykopping) [inline pkg/querier/queryrange/views.go:276]: Wrap the error please

File: pkg/querier/queryrange/views.go (hunk 8 above)
Comment (dannykopping) [reply pkg/querier/queryrange/views.go:1]: We could remove the whole directory just as easily though?

File: pkg/querier/queryrange/views.go (hunk 10 above)
Comment (jeschkies) [reply pkg/querier/queryrange/views.go]: Done. Although I'm not quite sure it is very useful in this case.

**Review by cstyan — COMMENTED** [review]
sorry, forgot to hit submit review on friday

File: pkg/querier/queryrange/views.go [hunk 11]
```diff
@@ -0,0 +1,250 @@
+package queryrange
+
+import (
+	"io"
+	"sort"
+
+	"github.com/cespare/xxhash/v2"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/richardartoul/molecule"
+	"github.com/richardartoul/molecule/src/codec"
```
[the remark this answers is not in the document]
Comment (jeschkies) [reply pkg/querier/queryrange/views.go:11]: Yeah, we seem to have to use the buffer.

File: pkg/querier/queryrange/views.go (hunk 6 above)
Comment (jeschkies) [reply pkg/querier/queryrange/views.go]: Done.

File: pkg/querier/queryrange/views.go (hunk 4 above)
Comment (jeschkies) [reply pkg/querier/queryrange/views.go]: Done

**Pull request merged**